### DISCUSSION TITLE
How do I call `request_verify_token` internally?

### DISCUSSION THREAD

**Discussion by ckmaresca** [fastapi-users/fastapi-users#654] [Q&A] [answered]
I'm using `on_after_register` to generate an email to the user for verification.   To make this easy for the user, I want to include a link for the user that has the token as a get argument.

To do this, I need to call `request_verify_token` to get the correct token.  I could do this via a JSON/REST call, but it seems stupid.  I would call `request_verify_token` directly, but I don't seem to be able to do that.

I'm probably missing something obvious....  Anyone have a hint on how to do this?

Thx.

Chris.

**Comment by ckmaresca** [comment, marked as answer]
I'm going to answer my own question.  It turns out the workflow is not what I expected and is unclear in the docs.   All of the docs suggest "send the user an email with the verification token".  I assumed that this was done in the `on_after_registration` hook, but that is not the case.  

The assumption is that this should be done in `after_verification_request` - my assumption was that `request_verify_token` would return a token, but it does not.  Instead it returns 202 with a NULL body, but calls `after_verification_request` with the token as the string `token`.

So, the correct workflow from the frontend is to first call `/register`, then call `/request_verify_token` - any logic to send email to a user needs to be in `after_verification_request`.   Also note that the verify token is a JWT, which makes it difficult to use SMS as a verification channel.

Anyway, all this is super unclear and the documentation for this is scattered all over the repo, issues & discusses.  Coupled with a minimalist example, it is a little challenging to implement.

HTH someone.

Chris. 

### References

- Verify Route Doc https://github.com/frankie567/fastapi-users/blob/61e5b7764760146927106c8e554b9b0858466c12/docs/configuration/routers/verify.md
- Generating Verify JWT https://github.com/frankie567/fastapi-users/discussions/604

**Comment by frankie567** [comment]
Hi Chris,

Thank you for your feedback. As you have pointed out, registration and verification logic are separated and have to be handled separetely. We provide building blocks that you are free to use or not. If you need special verification logic compatible with SMS, you can implement your own token generation/verification rather easily using `on_after_register` and a custom route.

If you think the documentation can be improved, I would be happy to accept a PR with more explanations and/or examples.
